Modules/socketmodule.c

```diff
@@ -896,9 +896,9 @@ sock_call_ex(PySocketSockObject *s,
         /* inner loop to retry sock_func() when sock_func() is interrupted
            by a signal */
         while (1) {
-            Py_BEGIN_ALLOW_THREADS
+            _Py_BEGIN_ALLOW_THREADS_COND(s->sock_timeout)
             res = sock_func(s, data);
-            Py_END_ALLOW_THREADS
+            _Py_END_ALLOW_THREADS_COND
```

Comment:
There are likely other socket operations where releasing the GIL could be avoided, but this gets the bulk of those done frequently in a loop (`send`/ `recv`, ...).   `socket.close()` is also handled below.